### PR TITLE
SUS-623: JSMessages deprecation: make JSMessages a wrapper around ResourceLoader

### PR DESCRIPTION
**Pull request by mszabo-wikia**
- `$.msg` now calls `mw.message(...).text()`
- `JSMessages::registerPackage` now registers a ResourceLoader module transparently
- `JSMessages::enqueuePackage` now selects this RL module for loading transparently

https://wikia-inc.atlassian.net/browse/SUS-623

@mixth-sense @macbre

### CODE REVIEW

File: extensions/wikia/JSMessages/JSMessages.class.php [hunk 1]
```diff
@@ -124,191 +81,142 @@ static private function resolveMessagesPattern($pattern) {
...
 	 */
-	static public function getPackages($packages, $allowWildcards = true) {
-		$messages = array();
-
-		foreach($packages as $packageName) {
-			$packageMessages = self::getPackage($packageName, $allowWildcards);
-
-			if (is_array($packageMessages)) {
-				$messages = array_merge($messages, $packageMessages);
+	static private function getMatchingMessagesForPattern( $pattern ) {
```
Comment (sqreek) [inline extensions/wikia/JSMessages/JSMessages.class.php:154]: should we check that `$pattern` ending with a `*`

File: extensions/wikia/JSMessages/JSMessages.class.php [hunk 2]
```diff
@@ -124,191 +81,142 @@ static private function resolveMessagesPattern($pattern) {
...
-
-		if ($url != "") {
-			// request a script
-			F::app()->wg->Out->addScript(Html::linkedScript($url));
+	static private function expandWildcards( $keys ) {
+		$list = [];
+		if ( is_array( $keys ) ) {
+			// Expand any wildcards, otherwise just push the message onto the list
+			foreach ( $keys as $message ) {
+				if ( substr( $message, -1 ) == '*' ) {
```
Comment (sqreek) [inline extensions/wikia/JSMessages/JSMessages.class.php]: If we move this conditional into `getMatchingMessagesForPatter`, and make it return $message when \* is not find it code will look nicer (less cyclometric)
In this case, method names should be changed for more proper one.

File: extensions/wikia/JSMessages/JSMessages.class.php [hunk 3]
```diff
@@ -124,191 +81,142 @@ static private function resolveMessagesPattern($pattern) {
...
+	 * @return bool true to continue hook processing
 	 */
-	static public function onResourceLoaderGetConfigVars(Array &$vars) {
-		$vars['wgJSMessagesCB'] = JSMessagesHelper::getMessagesCacheBuster();
+	static public function onResourceLoaderRegisterModules( ResourceLoader &$resourceLoader ) {
+		global $wgResourceModules;
+
+		foreach ( self::$packages as $packageName => $messageKeys ) {
+			$module = [
+				'messages' => self::expandWildcards( $messageKeys ),
```
Comment (sqreek) [inline extensions/wikia/JSMessages/JSMessages.class.php]: can we extract `messages` to static/const ?

File: extensions/wikia/JSMessages/JSMessages_setup.php [hunk 4]
```diff
@@ -9,34 +9,32 @@
...
 
 $wgExtensionFunctions[] = function () {
 	// This has to be wrapped in a function so it isn't run before we include GlobalSettings.php
-	JSMessages::registerPackage("ConfirmModal", array(
+	JSMessages::registerPackage( "ConfirmModal", [
 		'ok',
-		'cancel'
-	));
-	JSMessages::enqueuePackage("ConfirmModal", JSMessages::EXTERNAL); // We need this to ensure the messages are loaded on every page
+		'cancel',
```
Comment (sqreek) [inline extensions/wikia/JSMessages/JSMessages_setup.php]: unnecessary `,`

File: extensions/wikia/JSMessages/js/JSMessages.js [hunk 5]
```diff
@@ -72,36 +53,34 @@
...
+			mw.config.set('wgUserLanguageTemp', mw.config.get('wgUserLanguage'));
+			if (language) {
+				mw.config.set('wgUserLanguage', language);
 			}
 
-			// by default use user language
-			language = language || window.wgUserLanguage;
+			// append prefix
+			if ($.isArray(packages)) {
+				for (var i = 0; i < packages.length; i++) {
```
Comment (sqreek) [inline extensions/wikia/JSMessages/js/JSMessages.js]: `packages = packages.map(function(x){return modulePrefix + x;})`?

File: extensions/wikia/JSMessages/js/JSMessages.js [hunk 6]
```diff
@@ -113,125 +92,26 @@
...
 		JSMessages.getForContent = function(packages, callback) {
-			return JSMessages.get(packages, callback, window.wgContentLanguage);
+			return JSMessages.get(packages, callback, mw.config.get('wgContentLanguage'));
 		};
 
 		return JSMessages;
 	}
 
-	// private module for plural support
-	var Plurals = {
```
Comment (sqreek) [inline extensions/wikia/JSMessages/js/JSMessages.js:123]: Are we remove all Plurals support in JS? where it is handled now?

Comment (mszabo-wikia) [inline extensions/wikia/JSMessages/js/JSMessages.js:123]: It's provided by the core module `mediawiki.jqueryMsg`. Language data is loaded by ResourceLoader

Comment (sqreek) [inline extensions/wikia/JSMessages/js/JSMessages.js:123]: :+1:

**Comment by sqreek** [comment]
Let's make green light on test and go :+1:

**Pull request closed without merging**